File: src/tokenizer.c
#include "tokenizer.h"
#include <stdint.h>
#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>
#include <math.h>
#include <string.h>

#define TOKEN_POOL_SIZE 1000

void print_token(char* source, token token){
	char const* token_str[] = {
		"INT", "FLOAT",
		"STRING",
		"PLUS", "MINUS", "STAR", "SLASH", "NOT",
		"AMP", "AMP_AMP", "PIPE", "PIPE_PIPE",
		"EQUAL", "LESSER", "GREATER",
		"LESSER_LESSER", "GREATER_GREATER",
		"EQUAL_EQUAL", "NOT_EQUAL", "LESSER_EQUAL", "GREATER_EQUAL",
		"PLUS_EQUAL", "MINUS_EQUAL", "STAR_EQUAL", "SLASH_EQUAL",
		"COLON", "SEMICOLON", "DOT", "COMMA",
		"LEFT_PAREN", "RIGHT_PAREN", "LEFT_BRACE", "RIGHT_BRACE", "LEFT_BRACKET", "RIGHT_BRACKET",
		"IDENTIFIER",
		"LET", "CONST", "STRUCT", "FN", "IF", "ELSE", "FOR", "WHILE",
		"U8", "U16", "U32", "U64", "I8", "I16", "I32", "I64", "F32", "F64", "TRUE", "FALSE", "BOOL",
		"END_OF_FILE",
	};
	printf("Token, type: %s, str:", token_str[token.type]);
	for (uint32_t i = token.loc.start; i < token.loc.end; i++) {
			printf("%c", source[i]);
	}
	printf("\n");
}

char peek(char* source, scanner_status scan_status){
	return source[scan_status.current];
}

char peek_next(char* source, scanner_status scan_status){
	return source[scan_status.current + 1];
}

char consume(char* source, scanner_status* scan_status){
	return source[scan_status->current++];
}

bool match_consume(char* source, scanner_status* scan_status, char match){
	if(peek(source, *scan_status) != match) return false;
	consume(source, scan_status);
	return true;
}

void produce_token(token_pool* pool, token_type type, uint32_t loc_start, uint32_t loc_end){
	pool->pool[pool->cursor] = (token){.type = type, .loc.start = loc_start, .loc.end = loc_end};
	pool->cursor += 1;
}

bool is_digit(char c){
	return c >= '0' && c <= '9';
}

bool is_alpha(char c){
	return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z');
}

bool is_alpha_numeric(char c){
	return is_alpha(c) || is_digit(c);
}

uint32_t parse_number(char* source, scanner_status* scan_status, bool* is_float){
	while (is_digit(peek(source, *scan_status))) {
		consume(source, scan_status);
	}
	if(peek(source, *scan_status) == '.' && is_digit(peek_next(source, *scan_status))){
		consume(source, scan_status); // Consume .

		*is_float = true;
		while (is_digit(peek(source, *scan_status))) {
			consume(source, scan_status);
		}
	}
	return scan_status->current;
}

void report_error(char* source, uint32_t line, uint32_t at, const char* message){
	printf("Error at line: %d, %s\n", line, message);
	uint32_t error_line_offset = 0;
	uint32_t line_counter = 0;
	// Get offset to the line containing error.
	while (source[error_line_offset] != '\0') {
		if(line_counter == line) break;
		if(source[error_line_offset] == '\n') line_counter++;
		error_line_offset += 1;
	}
	// Print the line
	uint32_t i = error_line_offset;
	while (source[i] != '\n' && source[i] != '\0') {
		printf("%c", source[i]);
		i++;
	}
	printf("\n");
	// Print the fancy pointer
	for (uint32_t j = error_line_offset; j < i; j++) {
		if(j == at)
			printf("^");
		else
			printf("-");
	}
	printf("\n");
}

bool match_string(char* source, char* str1, uint32_t loc_start, uint32_t loc_end){
	uint32_t j = 0;
	if(strlen(str1) != (loc_end-loc_start)) return false;

	for (uint32_t i = loc_start; i < loc_end; i++) {
		if(str1[j] != source[i]) return false;
		j++;
	}
	return true;
}

uint32_t parse_identifier(char* source, scanner_status* scan_status, token_type* type){
	char* key_str[] = {
		"let", "const", "struct", "fn", "if", "else", "for", "while",
		"u8", "u16", "u32", "u64", "i8", "i16", "i32", "i64", "f32", "f64", "true", "false", "bool"
	};
	const token_type key_tok[] = {
		TOK_LET, TOK_CONST, TOK_STRUCT, TOK_FN, TOK_IF, TOK_ELSE, TOK_FOR, TOK_WHILE,
		TOK_U8, TOK_U16, TOK_U32, TOK_U64, TOK_I8, TOK_I16, TOK_I32, TOK_I64, TOK_F32, TOK_F64, TOK_TRUE, TOK_FALSE, TOK_BOOL
	};

	while (is_alpha_numeric(peek(source, *scan_status))) {
		consume(source, scan_status);
	}

	*type = TOK_IDENTIFIER;
	for (uint32_t i = 0; i < 21; i++) {
		bool found = match_string(source, key_str[i], scan_status->start, scan_status->current);
		if(found){
			*type = key_tok[i];
			break;
		}
	}
	return scan_status->current;
}

uint32_t parse_string(char* source, scanner_status* scan_status){
	uint32_t start = scan_status->start;
	while (peek(source, *scan_status) != '"' && peek(source, *scan_status) != '\0'){
		if(peek(source, *scan_status) == '\n') scan_status->line++;
		consume(source, scan_status);
	}
	if(peek(source, *scan_status) == '\0'){
		report_error(source, scan_status->line, start, "Unterminated String found");
	}
	// Consume the closing '"'.
	consume(source, scan_status);
	return scan_status->current;
}

token_pool scanner(char* source, uint32_t len){
	scanner_status scan_status = {};

	token_pool token_pool = {};
	token_pool.pool = calloc(TOKEN_POOL_SIZE, sizeof(token));

	while (scan_status.current < len) {
		scan_status.start = scan_status.current;
		char c = consume(source, &scan_status);
		switch (c) {
			case '+': {
				token_type type = match_consume(source, &scan_status, '=') ? TOK_PLUS_EQUAL : TOK_PLUS;
				produce_token(&token_pool, type, scan_status.start, scan_status.current);
				break;
			}
			case '-': {
				token_type type = match_consume(source, &scan_status, '=') ? TOK_MINUS_EQUAL : TOK_MINUS;
				produce_token(&token_pool, type, scan_status.start, scan_status.current);
				break;
			}
			case '*': {
				token_type type = match_consume(source, &scan_status, '=') ? TOK_STAR_EQUAL : TOK_STAR;
				produce_token(&token_pool, type, scan_status.start, scan_status.current);
				break;
			}
			case '/': {
				token_type type = match_consume(source, &scan_status, '=') ? TOK_SLASH_EQUAL : TOK_SLASH;
				if(type == TOK_SLASH && match_consume(source, &scan_status, '/')){
					while (peek(source, scan_status) != '\n' && scan_status.current < len) {
						consume(source, &scan_status);
					}
					break;
				}else{
					produce_token(&token_pool, type, scan_status.start, scan_status.current);
					break;
				}
				break;
			}
			case '=': {
				token_type type = match_consume(source, &scan_status, '=') ? TOK_EQUAL_EQUAL : TOK_EQUAL;
				produce_token(&token_pool, type, scan_status.start, scan_status.current);
				break;
			}
			case '<': {
				token_type type = match_consume(source, &scan_status, '=') ? TOK_LESSER_EQUAL : TOK_LESSER;
				if(type == TOK_LESSER) type = match_consume(source, &scan_status, '<') ? TOK_LESSER_LESSER : TOK_LESSER;
				produce_token(&token_pool, type, scan_status.start, scan_status.current);
				break;
			}
			case '>': {
				token_type type = match_consume(source, &scan_status, '=') ? TOK_GREATER_EQUAL : TOK_GREATER;
				if(type == TOK_GREATER) type = match_consume(source, &scan_status, '>') ? TOK_GREATER_GREATER : TOK_GREATER;
				produce_token(&token_pool, type, scan_status.start, scan_status.current);
				break;
			}
			case ':': produce_token(&token_pool, TOK_COLON, scan_status.start, scan_status.current); break;
			case '(': produce_token(&token_pool, TOK_LEFT_PAREN, scan_status.start, scan_status.current); break;
			case ')': produce_token(&token_pool, TOK_RIGHT_PAREN, scan_status.start, scan_status.current); break;
			case '{': produce_token(&token_pool, TOK_LEFT_BRACE, scan_status.start, scan_status.current); break;
			case '}': produce_token(&token_pool, TOK_RIGHT_BRACE, scan_status.start, scan_status.current); break;
			case '[': produce_token(&token_pool, TOK_LEFT_BRACKET, scan_status.start, scan_status.current); break;
			case ']': produce_token(&token_pool, TOK_RIGHT_BRACKET, scan_status.start, scan_status.current); break;
			case '.': produce_token(&token_pool, TOK_DOT, scan_status.start, scan_status.current); break;
			case ';': produce_token(&token_pool, TOK_SEMICOLON, scan_status.start, scan_status.current); break;
			case ',': produce_token(&token_pool, TOK_COMMA, scan_status.start, scan_status.current); break;
			case '"':{
				uint32_t loc_end = parse_string(source, &scan_status);
				produce_token(&token_pool, TOK_STRING, scan_status.start+1, loc_end-1);
				break;
			}
			case '!':{
				token_type type = match_consume(source, &scan_status, '=') ? TOK_NOT_EQUAL : TOK_NOT;
				produce_token(&token_pool, type, scan_status.start, scan_status.current); break;
			}
			case '&':{
				token_type type = match_consume(source, &scan_status, '&') ? TOK_AMP_AMP : TOK_AMP;
				produce_token(&token_pool, type, scan_status.start, scan_status.current); break;
			}
			case '|':{
				token_type type = match_consume(source, &scan_status, '|') ? TOK_PIPE_PIPE : TOK_PIPE;
				produce_token(&token_pool, type, scan_status.start, scan_status.current); break;
			}
			case ' ': break;
			case '\r': break;
			case '\t': break;
			case '\n': scan_status.line++; break;
			default:
				if(is_digit(c)){
					bool is_float = false;
					uint32_t toc_end = parse_number(source, &scan_status, &is_float);
					if(is_float){
						produce_token(&token_pool, TOK_FLOAT, scan_status.start, toc_end);
					}else{
						produce_token(&token_pool, TOK_INT, scan_status.start, toc_end);
					}
				}else if(is_alpha(c)){
					token_type type;
					uint32_t toc_end = parse_identifier(source, &scan_status, &type);
					produce_token(&token_pool, type, scan_status.start, toc_end);
				}
				else{
					report_error(source, scan_status.line, scan_status.start, "Unexpected char found.");
				}
		}
	}
	produce_token(&token_pool, TOK_END_OF_FILE, scan_status.start, 0);
	return token_pool;
}

File: src/tokenizer.zig
const std = @import("std");

pub const LocInfo = struct {
    start: u32,
    end: u32,
    line: u32,
};

pub const TokenType = enum {
    tok_int,
    tok_float,
    tok_plus,
    tok_minus,
    tok_star,
    tok_slash,
    tok_not,
    tok_amp,
    tok_amp_amp,
    tok_pipe,
    tok_pipe_pipe,
    tok_equal,
    tok_equal_equal,
    tok_not_equal,
    tok_greater_equal,
    tok_lesser_equal,
    tok_greater,
    tok_lesser,
    tok_left_paren,
    tok_right_paren,
    tok_left_brace,
    tok_right_brace,
    tok_colon,
    tok_semi_colon,
    tok_true,
    tok_false,
    tok_bool,
    tok_identifier,
    tok_var,
    tok_if,
    tok_else,
    tok_print,
    tok_int_type,
    tok_float_type,
    tok_bool_type,
    tok_fn,
    tok_void_type,
    tok_return,
    tok_comma,
    tok_eof,

    pub fn str(token_type: TokenType) []const u8 {
        switch (token_type) {
            .tok_int => return "int",
            .tok_float => return "float",
            .tok_plus => return "plus",
            .tok_minus => return "minus",
            .tok_star => return "star",
            .tok_slash => return "slash",
            .tok_not => return "not",
            .tok_amp => return "and",
            .tok_amp_amp => return "amp_amp",
            .tok_pipe => return "pipe",
            .tok_pipe_pipe => return "pipe_pipe",
            .tok_equal => return "equal",
            .tok_equal_equal => return "equal_equal",
            .tok_not_equal => return "not_equal",
            .tok_greater_equal => return "greater_equal",
            .tok_lesser_equal => return "lesser_equal",
            .tok_greater => return "greater",
            .tok_lesser => return "lesser",
            .tok_left_paren => return "left_paren",
            .tok_right_paren => return "right_paren",
            .tok_left_brace => return "left_brace",
            .tok_right_brace => return "right_brace",
            .tok_colon => return "colon",
            .tok_semi_colon => return "semi_colon",
            .tok_true => return "true",
            .tok_false => return "false",
            .tok_bool => return "bool",
            .tok_identifier => return "identifier",
            .tok_var => return "var",
            .tok_if => return "if",
            .tok_else => return "else",
            .tok_print => return "print",
            .tok_int_type => return "int_type",
            .tok_float_type => return "float_type",
            .tok_bool_type => return "bool_type",
            .tok_fn => return "fn",
            .tok_void_type => return "void_type",
            .tok_return => return "return",
            .tok_comma => return "comma",
            .tok_eof => return "eof",
        }
    }

    pub fn reservedStr(token_type: TokenType) []const u8 {
        switch (token_type) {
            .tok_var => return "var",
            .tok_if => return "if",
            .tok_else => return "else",
            .tok_print => return "print",
            .tok_int_type => return "int",
            .tok_float_type => return "float",
            .tok_bool_type => return "bool",
            .tok_true => return "true",
            .tok_false => return "false",
            .tok_fn => return "fn",
            .tok_void_type => return "void",
            .tok_return => return "return",
            else => {
                return "";
            },
        }
    }
};

pub const Token = struct {
    type: TokenType,
    loc: LocInfo,
};

pub const Tokenizer = struct {
    source: []u8,
    source_name: []const u8,
    start: u32 = 0,
    current: u32 = 0,
    line: u32 = 0,
    pool: std.ArrayList(Token),

    pub fn init(allocator: std.mem.Allocator, source: []u8, source_name: []const u8) Tokenizer {
        const tokenizer: Tokenizer = .{
            .source = source,
            .source_name = source_name,
            .pool = std.ArrayList(Token).init(allocator),
        };
        return tokenizer;
    }

    fn peek(tokenizer: *Tokenizer) u8 {
        if (tokenizer.current >= tokenizer.source.len) return 0;
        return tokenizer.source[tokenizer.current];
    }

    fn peek_next(tokenizer: *Tokenizer) u8 {
        if (tokenizer.current >= tokenizer.source.len) return 0;
        return tokenizer.source[tokenizer.current + 1];
    }

    fn consume(tokenizer: *Tokenizer) u8 {
        const char = tokenizer.peek();
        tokenizer.current += 1;
        return char;
    }

    fn match(tokenizer: *Tokenizer, char: u8) bool {
        if (tokenizer.peek() != char) return false;
        _ = tokenizer.consume();
        return true;
    }

    fn add_token(tokenizer: *Tokenizer, token_type: TokenType) void {
        tokenizer.pool.append(.{ .type = token_type, .loc = .{ .start = tokenizer.start, .end = tokenizer.current, .line = tokenizer.line } }) catch |err| {
            std.debug.print("Error while adding token: {any}\n", .{err});
        };
    }

    fn parseIdentifier(tokenizer: *Tokenizer) TokenType {
        const reserved = [_]TokenType{ .tok_true, .tok_false, .tok_var, .tok_print, .tok_int_type, .tok_float_type, .tok_bool_type, .tok_fn, .tok_void_type, .tok_return, .tok_if, .tok_else };

        var out_token_type: TokenType = .tok_identifier;

        while (tokenizer.current < tokenizer.source.len and std.ascii.isAlphanumeric(tokenizer.peek())) _ = tokenizer.consume();

        for (reserved, 0..) |token_type, i| {
            if (std.mem.eql(u8, tokenizer.source[tokenizer.start..tokenizer.current], token_type.reservedStr())) {
                out_token_type = reserved[i];
                break;
            }
        }

        return out_token_type;
    }

    fn parseNumber(tokenizer: *Tokenizer) bool {
        var is_float = false;
        while (tokenizer.current < tokenizer.source.len and std.ascii.isDigit(tokenizer.peek())) _ = tokenizer.consume();
        if (tokenizer.peek() == '.' and std.ascii.isDigit(tokenizer.peek_next())) {
            _ = tokenizer.consume();
            is_float = true;
            while (tokenizer.current < tokenizer.source.len and std.ascii.isDigit(tokenizer.peek())) _ = tokenizer.consume();
        }
        return is_float;
    }

    pub fn tokenize(tokenizer: *Tokenizer) void {
        while (tokenizer.current < tokenizer.source.len) {
            tokenizer.start = tokenizer.current;
            const char: u8 = tokenizer.consume();
            switch (char) {
                '+' => {
                    tokenizer.add_token(.tok_plus);
                },
                '-' => {
                    tokenizer.add_token(.tok_minus);
                },
                '*' => {
                    tokenizer.add_token(.tok_star);
                },
                '/' => {
                    if (tokenizer.match('/')) {
                        while (tokenizer.current < tokenizer.source.len and tokenizer.consume() != '\n') {}
                        tokenizer.line += 1;
                    } else {
                        tokenizer.add_token(.tok_slash);
                    }
                },
                '=' => {
                    const token_type: TokenType = if (tokenizer.match('=')) .tok_equal_equal else .tok_equal;
                    tokenizer.add_token(token_type);
                },
                '!' => {
                    const token_type: TokenType = if (tokenizer.match('=')) .tok_not_equal else .tok_not;
                    tokenizer.add_token(token_type);
                },
                '&' => {
                    const token_type: TokenType = if (tokenizer.match('&')) .tok_amp_amp else .tok_amp;
                    tokenizer.add_token(token_type);
                },
                '|' => {
                    const token_type: TokenType = if (tokenizer.match('|')) .tok_pipe_pipe else .tok_pipe;
                    tokenizer.add_token(token_type);
                },
                '>' => {
                    const token_type: TokenType = if (tokenizer.match('=')) .tok_greater_equal else .tok_greater;
                    tokenizer.add_token(token_type);
                },
                '<' => {
                    const token_type: TokenType = if (tokenizer.match('=')) .tok_lesser_equal else .tok_lesser;
                    tokenizer.add_token(token_type);
                },
                '(' => tokenizer.add_token(.tok_left_paren),
                ')' => tokenizer.add_token(.tok_right_paren),
                '{' => tokenizer.add_token(.tok_left_brace),
                '}' => tokenizer.add_token(.tok_right_brace),
                ':' => tokenizer.add_token(.tok_colon),
                ';' => tokenizer.add_token(.tok_semi_colon),
                ',' => tokenizer.add_token(.tok_comma),
                ' ' => {},
                '\r' => {},
                '\t' => {},
                '\n' => tokenizer.line += 1,
                else => {
                    if (std.ascii.isDigit(char)) {
                        const is_float = tokenizer.parseNumber();
                        tokenizer.add_token(if (is_float) .tok_float else .tok_int);
                    } else if (std.ascii.isAlphabetic(char)) {
                        const token_type = tokenizer.parseIdentifier();
                        tokenizer.add_token(token_type);
                    } else {
                        tokenizer.reportError("Unexpected char found!", true);
                    }
                },
            }
        }
        tokenizer.add_token(.tok_eof);
    }

    fn reportError(tokenizer: *Tokenizer, err: []const u8, exit: bool) void {
        std.debug.print("Error at line {d}: {s}\n", .{ tokenizer.line, err });

        var error_line_offset: u32 = 0;
        var line_counter: u32 = 0;
        // Get offset to the line containing error.
        while (error_line_offset < tokenizer.source.len) : (error_line_offset += 1) {
            if (line_counter == tokenizer.line) break;
            if (tokenizer.source[error_line_offset] == '\n') line_counter += 1;
        }
        // Print the line
        var i: u32 = error_line_offset;
        while (i < tokenizer.source.len and tokenizer.source[i] != '\n') : (i += 1) {
            std.debug.print("{c}", .{tokenizer.source[i]});
        }
        std.debug.print("\n", .{});
        // Print the fancy pointer
        for (error_line_offset..i) |j| {
            if (j == tokenizer.start) std.debug.print("^", .{}) else std.debug.print("-", .{});
        }
        std.debug.print("\n", .{});

        std.process.exit(@intFromBool(exit));
    }

    pub fn print(tokenizer: *Tokenizer) void {
        for (0..tokenizer.pool.items.len) |i| {
            const token = tokenizer.pool.items[i];
            std.debug.print("Token, type: {s}, src: {s}\n", .{ token.type.str(), tokenizer.source[token.loc.start..token.loc.end] });
        }
    }

    pub fn deinit(tokenizer: *Tokenizer) void {
        tokenizer.pool.deinit();
    }
};
